File: sub-web-service/src/main/java/org/github/web/module/sys/menu/service/impl/MenuServiceImpl.java
package org.github.web.module.sys.menu.service.impl;

import lombok.NonNull;
import lombok.val;
import org.github.base.Page;
import org.github.cache.RAMCache;
import org.github.mysql.web.base.entity.SysMenuMbpEntity;
import org.github.mysql.web.base.service.ISysMenuMbpService;
import org.github.spring.restful.json.JSONArrayReturn;
import org.github.spring.restful.json.JSONDataReturn;
import org.github.spring.restful.json.JSONPageReturn;
import org.github.web.module.sys.menu.model.bo.QueryMenuListBO;
import org.github.web.module.sys.menu.model.vo.QueryMenuListVO;
import org.github.web.module.sys.menu.service.IMenuService;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.util.List;

import static cn.hutool.core.text.CharSequenceUtil.isNotBlank;
import static org.github.mysql.web.base.enums.MenuType.button;
import static org.github.mysql.web.base.enums.MenuType.directory;
import static org.github.web.common.CacheName.*;
import static org.springframework.beans.BeanUtils.copyProperties;

@Service
public class MenuServiceImpl implements IMenuService {
  @Autowired
  private ISysMenuMbpService SysMenuMbpService;
  @Autowired
  private RAMCache           ramCache;

  @Override
  public @NonNull JSONDataReturn<Boolean> delMenuList(@NonNull List<Long> ids) {
    if (ids.isEmpty()) return JSONDataReturn.of(false);
    val result = SysMenuMbpService.removeByIds(ids);
    return JSONDataReturn.of(result);
  }

  @Override
  public @NonNull JSONDataReturn<SysMenuMbpEntity> queryMenuInfo(@NonNull Long menuId) {
    val entity = SysMenuMbpService.getById(menuId);
    return JSONDataReturn.of(entity);
  }

  @Override
  public @NonNull JSONPageReturn<QueryMenuListVO> queryMenuPage(@NonNull QueryMenuListBO bo) {
    val query = SysMenuMbpService.lambdaQuery();
    query.likeRight(isNotBlank(bo.getName()), SysMenuMbpEntity::getName, bo.getName());
    val page = query.page(new Page<>(bo));
    val list = page.getRecords();
    if (list.isEmpty()) return JSONPageReturn.of(page);
    return JSONPageReturn.of(page, this::applySysMenuMbpEntity2QueryMenuListVO);
  }

  @Override
  public @NonNull JSONArrayReturn<SysMenuMbpEntity> queryMenuTree() {
    val query = SysMenuMbpService.lambdaQuery();
    query.ne(SysMenuMbpEntity::getType, button.getCode());
    val list = query.list();

    val root = new SysMenuMbpEntity();
    root.setMenuId(0L);
    root.setParentId(-1L);
    root.setName("根目录");
    // root.setUrl();
    // root.setPerms();
    root.setType(directory.getCode());
    root.setIcon("fa fa-car");
    root.setOrderNum(0);
    list.add(root);

    return JSONArrayReturn.of(list);
  }

  @Override
  public @NonNull JSONDataReturn<Boolean> saveOrUpdate(@NonNull SysMenuMbpEntity bo) {
    val result = SysMenuMbpService.saveOrUpdate(bo);
    return JSONDataReturn.of(result);
  }

  private @NonNull QueryMenuListVO applySysMenuMbpEntity2QueryMenuListVO(@NonNull SysMenuMbpEntity entity) {
    val vo = new QueryMenuListVO();
    copyProperties(entity, vo);
    vo.setParentName(ramCache.<String>getCache(sysMenuName).get(vo.getParentId()));
    return vo;
  }
}

File: sub-web-service/src/main/java/org/github/web/module/sys/menu/service/impl/MenuServiceImpl.kt
package org.github.web.module.sys.menu.service.impl

import org.github.web.module.sys.menu.service.IMenuService
import org.springframework.beans.factory.annotation.Autowired
import org.github.mysql.web.base.service.ISysMenuService
import org.github.spring.restful.json.JSONDataReturn
import org.github.mysql.web.base.entity.SysMenuEntity
import org.github.web.module.sys.menu.model.bo.QueryMenuListBO
import org.github.spring.restful.json.JSONPageReturn
import org.github.web.module.sys.menu.model.vo.QueryMenuListVO
import org.github.base.Page
import org.github.mybatis.ops.ktGetById
import org.github.mybatis.ops.ktList
import org.github.mybatis.ops.ktQueryWrapper
import org.github.spring.restful.json.JSONArrayReturn
import org.github.mysql.web.base.dict.MenuType.BUTTON
import org.github.mysql.web.base.dict.MenuType.DIRECTORY
import org.github.web.module.sys.menu.mapper.IMenuMapper
import org.springframework.stereotype.Service

@Service
class MenuServiceImpl: IMenuService {
  @Autowired
  private lateinit var sysMenuService: ISysMenuService
  @Autowired
  private lateinit var menuMapper: IMenuMapper

  override fun delMenuList(ids: List<Long>): JSONDataReturn<Boolean> {
    if (ids.isEmpty()) return JSONDataReturn.of(false)
    val result = sysMenuService.removeByIds(ids)
    return JSONDataReturn.of(result)
  }

  override fun queryMenuInfo(menuId: Long): JSONDataReturn<SysMenuEntity?> {
    val entity = sysMenuService.ktGetById(menuId)
    return JSONDataReturn.of(entity)
  }

  override fun queryMenuPage(bo: QueryMenuListBO): JSONPageReturn<QueryMenuListVO> {
    val page = menuMapper.queryPage(Page(bo), bo)
    return JSONPageReturn.of(page)
  }

  override fun queryMenuTree(): JSONArrayReturn<SysMenuEntity> {
    val query = sysMenuService.ktQueryWrapper()
    query.ne(SysMenuEntity::type, BUTTON.code)
    val list = query.ktList()
    val root = SysMenuEntity()
    root.menuId = 0L
    root.parentId = -1L
    root.name = "根目录"
    // root.setUrl();
    // root.setPerms();
    root.type = DIRECTORY.code
    root.icon = "fa fa-car"
    root.orderNum = 0
    list.add(root)
    return JSONArrayReturn.of(list)
  }

  override fun saveOrUpdate(bo: SysMenuEntity): JSONDataReturn<Boolean> {
    val result = sysMenuService.saveOrUpdate(bo)
    return JSONDataReturn.of(result)
  }
}
